tools/please_go/packageinfo/packageinfo.go

```diff
@@ -0,0 +1,130 @@
+// Package packageinfo writes information about Go packages in a JSON format.
+// This is created at build time and intended to be consumed by the gopackagedriver binary.
+package packageinfo
+
+import (
+	"encoding/json"
+	"fmt"
+	"go/parser"
+	"go/scanner"
+	"go/token"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+
+	"github.com/please-build/go-rules/tools/please_go/embed"
+)
+
+// WritePackageInfo writes package info to the given file.
+func WritePackageInfo(importPath, plzPkg string, goSrcs []string, embedCfg string, w io.Writer) error {
+	pkg, err := createPackage(filepath.Join(importPath, plzPkg), goSrcs, embedCfg)
+	if err != nil {
+		return err
+	}
+	return serialise([]*packages.Package{pkg}, w)
+}
+
+// WriteModuleInfo writes a series of package info files to the given file.
+func WriteModuleInfo(modulePath, strip, src string, w io.Writer) error {
+	// Discover all Go files in the module
+	goFiles := map[string][]string{}
+	if err := filepath.WalkDir(src, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		} else if name := d.Name(); name == "testdata" {
+			return filepath.SkipDir // Don't descend into testdata
+		} else if strings.HasSuffix(name, ".go") && !strings.HasSuffix(name, "_test.go") {
+			dir := filepath.Dir(path)
+			goFiles[dir] = append(goFiles[dir], path)
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to read module dir: %w", err)
+	}
+	pkgs := make([]*packages.Package, 0, len(goFiles))
+	for dir, files := range goFiles {
+		pkgDir := strings.TrimPrefix(strings.TrimPrefix(dir, strip), "/")
+		pkg, err := createPackage(filepath.Join(modulePath, pkgDir), files, "")
+		if err != nil {
+			return err
+		}
+		pkgs = append(pkgs, pkg)
+	}
+	// Ensure output is deterministic
+	sort.Slice(pkgs, func(i, j int) bool {
+		return pkgs[i].ID < pkgs[j].ID
+	})
+	return serialise(pkgs, w)
+}
+
+func createPackage(pkgPath string, goSrcs []string, embedCfg string) (*packages.Package, error) {
```

Comment:
Sorry if I'm missing something, but I'm having trouble following this. Can't we just use https://pkg.go.dev/go/build#ImportDir, and let it populate the package for us?

Reply:
Hmmm I think we can, although I also don't think it will simplify much of this - the two structures are both called `Package` but they aren't really as similar as all that :(

Reply:
Okay was actually better than I thought since it deals with embed stuff and imports too. Doesn't solve _everything_ because of the myriad `Package` types but defo helps.

Reply:
oh yep and better than I thought - the AST stuff isn't actually serialised anyway...